extensions/sql-database-projects/src/test/projectController.test.ts

```diff
@@ -27,43 +30,99 @@ describe('ProjectsController: project controller operations', function (): void
-		should(project.dataSources.length).equal(2); // detailed datasources tests in their own test file
+				const projController = new ProjectsController(testContext.apiWrapper.object, new SqlDatabaseProjectTreeViewProvider());
+				const project = new Project('FakePath');
+
+				should(project.files.length).equal(0);
+				await projController.addItemPrompt(new Project('FakePath'), '', templates.script);
+				should(project.files.length).equal(0, 'Expected to return without throwing an exception or adding a file when an empty/undefined name is provided.');
+			}
+		});
 	});
 
-	it('Should return silently when no object name provided', async function (): Promise<void> {
-		for (const name of ['', '    ', undefined]) {
-			testContext.apiWrapper.reset();
-			testContext.apiWrapper.setup(x => x.showInputBox(TypeMoq.It.isAny(), TypeMoq.It.isAny())).returns(() => Promise.resolve(name));
-			testContext.apiWrapper.setup(x => x.showErrorMessage(TypeMoq.It.isAny())).returns((s) => { console.log('we throwin'); throw new Error(s); });
+	describe('Deployment and deployment script generation', function (): void {
+		it('Deploy dialog should open from ProjectController', async function (): Promise<void> {
+			let opened = false;
 
-			const projController = new ProjectsController(testContext.apiWrapper.object, new SqlDatabaseProjectTreeViewProvider());
-			const project = new Project('FakePath');
+			let deployDialog = TypeMoq.Mock.ofType(DeployDatabaseDialog);
+			deployDialog.setup(x => x.openDialog()).returns(() => { opened = true; });
+
+			let projController = TypeMoq.Mock.ofType(ProjectsController);
+			projController.callBase = true;
+			projController.setup(x => x.getDeployDialog(TypeMoq.It.isAny())).returns(() => deployDialog.object);
+
+			await projController.object.deployProject(new Project('FakePath'));
+			should(opened).equal(true);
+		});
+
+		it('Callbacks are hooked up and called from Deploy dialog', async function (): Promise<void> {
+			const projPath = path.dirname(await testUtils.createTestSqlProjFile(baselines.openProjectFileBaseline));
+			await testUtils.createTestDataSources(baselines.openDataSourcesBaseline, projPath);
+			const proj = new Project(projPath);
+
+			const deployHoller = 'hello from callback for deploy()';
+			const generateHoller = 'hello from callback for generateScript()';
+
+			let holler = 'nothing';
+
+			let deployDialog = TypeMoq.Mock.ofType(DeployDatabaseDialog, undefined, undefined, new ApiWrapper(), proj);
+			deployDialog.callBase = true;
+			deployDialog.setup(x => x.getConnectionUri()).returns(async () => 'fake|connection|uri');
+
+			let projController = TypeMoq.Mock.ofType(ProjectsController);
+			projController.callBase = true;
+			projController.setup(x => x.getDeployDialog(TypeMoq.It.isAny())).returns(() => deployDialog.object);
+			projController.setup(x => x.executionCallback(TypeMoq.It.isAny(), TypeMoq.It.is((_): _ is IDeploymentProfile => true))).returns(async () => {
```

Comment:
TypeMoq.It.is((_): 

what does this mean?

Reply:
It's how you provide a lambda function as a conditional for that argument.  In this case, I'm having the mock perform different actions (setting `holler` to different values) depending on whether the second argument is an `IDeploymentProfile` (this line) or an `IGenerateScriptProfile` (line 112).

I could have also done that switching logic within the mock function itself, but doing it in the mock setup seemed more succinct.

Reply:
The `_` indicates to typescript that it's a discard, so it shouldn't complain about me not using that variable in the definition of the function.